barretenberg/cpp/src/barretenberg/stdlib/primitives/databus/databus.hpp

```diff
@@ -77,54 +79,95 @@ template <class Builder> class DataBusDepot {
-            // Previous kernel return data should correspond to the calldata of the subsequent kernel
-            assert_equality_of_commitments(kernel_return_data, commitments.calldata);
+            start_idx = propagation_data.app_return_data_public_input_idx;
+            Commitment app_return_data = reconstruct_commitment_from_public_inputs(public_inputs, start_idx);
+
+            // Assert equality between the corresponding calldata and return data commitments
+            assert_equality_of_commitments(kernel_return_data, calldata);
+            assert_equality_of_commitments(app_return_data, secondary_calldata);
         }
+    }
 
-        // Propagate the return data commitment via the public inputs mechanism
-        propagate_commitment_via_public_inputs(commitments.return_data, is_kernel_data);
-    };
+    /**
+     * @brief Propagate the existing return data commitments via the public inputs of the provided circuit
+     * @details For consistent behavior across kernels, every kernel propagates two return data commitments via its
+     * public inputs. If one of either the app or kernel return data does not exist, it is populated with a default
+     * value that will satisfy the consistency check on the next cycle. For example, the first kernel has no previous
+     * kernel to verify and thus neither receives a previous kernel return data commitment nor a calldata input
+     * corresponding to a previous kernel. The commitment to the "empty" calldata will take a default value and thus we
+     * set the same value for the missing return data so that the consistency check will be satisfied. See issue
+     * https://github.com/AztecProtocol/barretenberg/issues/1138 for more discussion.
+     * @note The ordering of the kernel/app return data commitments within the public inputs is arbitrary but must be
+     * consistent across all kernels in order for the corresponding conistency check constraints to be consistent.
+     *
+     * @param builder
+     */
+    void propagate_return_data_commitments(Builder& builder)
+    {
+        // Set default commitment value to be used in the absence of one or the other return_data commitment
+        CommitmentNative default_commitment_val = CommitmentNative::one() * FrNative(BusVector::DEFAULT_VALUE);
+        if (kernel_return_data_commitment_exists) {
+            propagate_commitment_via_public_inputs(kernel_return_data_commitment, /*is_kernel=*/true);
+        } else {
+            Commitment default_commitment = Commitment::from_witness(&builder, default_commitment_val);
+            propagate_commitment_via_public_inputs(default_commitment, /*is_kernel=*/true);
+        }
+
+        if (app_return_data_commitment_exists) {
+            propagate_commitment_via_public_inputs(app_return_data_commitment, /*is_kernel=*/false);
+        } else {
+            Commitment default_commitment = Commitment::from_witness(&builder, default_commitment_val);
```

Comment:
I think we shouldn't use `from_witness` here since the default_commitment_val is a constant (also above)

Reply:
Added some logic for converting constants to fixed witnesses as we discussed